### ISSUE TITLE
promscale performance issue

### ISSUE THREAD

**Issue by brasno** [timescale/promscale#1214]
Dear people,
we have recently run into promscale/timescale performance issues. 
We are running Postgres-13 leader/replica cluster (Patroni) with timescale/promscale extensions.  Few days ago we upgraded Promscale from version 0.52. to to latest promscale version (Version:  0.10.0 Commit Hash:  0742e64f1c1371f6d83fa761f45b54dc4e69e5a8) that was available and since then we are not able to get system to perform as it did before the upgrade. We believe that there might be some issues with promscale or timescale/promscale_extension that is causing this.
List of active **extensions on timescale instance**:
timescale=# select  oid, extname, extowner, extnamespace, extrelocatable, extversion  from pg_extension;
```
   oid    |   extname   | extowner | extnamespace | extrelocatable | extversion 
---------------------------------------------------------------------------------
    13420 | plpgsql     |       10 |           11 | f              | 1.0        |
 64102680 | promscale   |       10 |      6566252 | f              | 0.3.0      |
  6565677 | timescaledb |       10 |         2200 | f              | 2.2.1      |
---------------------------------------------------------------------------------
```
**Default settings in database:**
timescale=# select * from _prom_catalog.default;
```
          key           |  value
------------------------+----------
chunk_interval         | 08:00:00
metric_compression     | true
ha_lease_timeout       | 1m
ha_lease_refresh       | 10s
trace_retention_period | 30 days
retention_period       | 31 days
```
**Database size**
```
timescale=# SELECT pg_size_pretty( pg_database_size('timescale') );
pg_size_pretty
----------------
487 GB
```
Two servers with 32 cores/192GB RAM are used for Postgres/timescale.
There are two instances of promscale which handle promteheus remote writes:
**/etc/sysconfig/promscale.conf**
```
PROMSCALE_DB_HOST="192.168.250.138"
PROMSCALE_DB_NAME="timescale"
PROMSCALE_DB_PASSWORD="***"
PROMSCALE_DB_USER="****"
PROMSCALE_DB_PORT="5000"
PROMSCALE_DB_SSL_MODE="disable"
PROMSCALE_DB_CONNECTION_TIMEOUT="5m0s"
PROMSCALE_DB_WRITER_CONNECTION_CONCURRENCY="12"
OPTIONS=" -metrics.high-availability=true -web.enable-admin-api -metrics.ignore-samples-written-to-compressed-chunks=true -log-level=info -metrics.async-acks=true"
PROMSCALE_PKG="rpm"

```
We added these CLI parameters in OPTIONS beside existing: -metrics.high-availability=true parameter but no change in performance was observed.

**Promscale startup:**
```
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.760Z caller=flags.go:156 msg="Using deprecated flags that will be dropped in a future version. Please update to new flag names."
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.761Z caller=flags.go:158 msg="Update deprecated flag db-connection-timeout to new flag db.connection-timeout"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.761Z caller=flags.go:158 msg="Update deprecated flag db-host to new flag db.host"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.761Z caller=flags.go:158 msg="Update deprecated flag db-user to new flag db.user"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.761Z caller=flags.go:158 msg="Update deprecated flag db-writer-connection-concurrency to new flag db.num-writer-connections"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.761Z caller=flags.go:158 msg="Update deprecated flag log-level to new flag telemetry.log.level"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.761Z caller=flags.go:158 msg="Update deprecated flag db-name to new flag db.name"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.761Z caller=flags.go:158 msg="Update deprecated flag db-password to new flag db.password"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.761Z caller=flags.go:158 msg="Update deprecated flag db-port to new flag db.port"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.761Z caller=flags.go:158 msg="Update deprecated flag db-ssl-mode to new flag db.ssl-mode"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:10.762Z caller=runner.go:60 msg="Version:0.10.0; Commit Hash: 0742e64f1c1371f6d83fa761f45b54dc4e69e5a8"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:10.763Z caller=runner.go:65 config="{ListenAddr::9201 ThanosStoreAPIListenAddr: OTLPGRPCListenAddr: PgmodelCfg:{CacheConfig:{SeriesCacheInitialSize:250000 seriesCacheMemoryMaxFlag:{kind:0 value:50} SeriesCacheMemoryMaxBytes:26896154624 MetricsCacheSize:10000 LabelsCacheSize:10000 ExemplarKeyPosCacheSize:10000} AppName:promscale@0.10.0 Host:192.168.250.138 Port:5000 User:postgres Password:**** Database:timescale SslMode:disable DbConnectionTimeout:5m0s IgnoreCompressedChunks:true AsyncAcks:true WriteConnectionsPerProc:12 MaxConnections:-1 UsesHA:true DbUri:**** EnableStatementsCache:true} LogCfg:{Level:info Format:logfmt} TracerCfg:{OtelCollectorEndpoint: OtelCollectorTLSCertFile: OtelCollectorTLSKeyFile: JaegerCollectorEndpoint: SamplingRatio:1} APICfg:{AllowedOrigin:^(?:.*)$ ReadOnly:false HighAvailability:true AdminAPIEnabled:true TelemetryPath:/metrics Auth:0xc000145e00 MultiTenancy:<nil> EnabledFeatureMap:map[] PromscaleEnabledFeatureList:[] MaxQueryTimeout:2m0s SubQueryStepInterval:1m0s LookBackDelta:5m0s MaxSamples:50000000 MaxPointsPerTs:11000} LimitsCfg:{targetMemoryFlag:{kind:0 value:80} TargetMemoryBytes:53792309248} TenancyCfg:{SkipTenantValidation:false EnableMultiTenancy:false AllowNonMTWrites:false ValidTenantsStr:allow-all ValidTenantsList:[]} ConfigFile:config.yml DatasetConfig: TLSCertFile: TLSKeyFile: ThroughputInterval:1s Migrate:true StopAfterMigrate:false UseVersionLease:true InstallExtensions:true UpgradeExtensions:true UpgradePrereleaseExtensions:false StartupOnly:false}"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.848Z caller=config.go:192 msg="had to reduce the number of copiers due to connection limits: wanted 96, reduced to 50"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:10.849Z caller=client.go:115 msg="postgresql://postgres:%2A%2A%2A%2A@192.168.250.138:5000/timescale?application_name=promscale%400.10.0&connect_timeout=300&sslmode=disable" numCopiers=92 pool_max_conns=100 pool_min_conns=8 statement_cache="512 statements"
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=warn ts=2022-03-02T19:43:10.884Z caller=telemetry.go:56 msg="Promscale collects anonymous usage telemetry data to help the Promscale team better understand and assist users. This can be disabled via the process described at https://docs.timescale.com/timescaledb/latest/how-to-guides/configuration/telemetry/#disabling-telemetry
"Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:10.917Z caller=runner.go:122 msg="Starting up..."
Mar  2 20:43:10 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:10.917Z caller=runner.go:123 msg=Listening addr=:9201
Mar  2 20:43:12 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:12.763Z caller=throughput.go:76 msg="ingestor throughput" samples/sec=0 metrics-max-sent-ts=1970-01-01T00:00:00Z
Mar  2 20:43:14 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:14.763Z caller=throughput.go:76 msg="ingestor throughput" samples/sec=0 metrics-max-sent-ts=1970-01-01T00:00:00Z
Mar  2 20:43:15 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:15.763Z caller=throughput.go:76 msg="ingestor throughput" samples/sec=0 metrics-max-sent-ts=1970-01-01T00:00:00Z
Mar  2 20:43:18 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:18.763Z caller=throughput.go:76 msg="ingestor throughput" samples/sec=0 metrics-max-sent-ts=1970-01-01T00:00:00Z
Mar  2 20:43:51 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:51.764Z caller=throughput.go:76 msg="ingestor throughput" samples/sec=0 metrics-max-sent-ts=1970-01-01T00:00:00Z
Mar  2 20:43:52 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:52.764Z caller=throughput.go:76 msg="ingestor throughput" samples/sec=0 metrics-max-sent-ts=1970-01-01T00:00:00Z
Mar  2 20:43:59 lxnmspromscalev1p promscale[72345]: level=info ts=2022-03-02T19:43:59.769Z caller=throughput.go:76 msg="ingestor throughput" samples/sec=0 metrics-max-sent-ts=1970-01-01T00:00:00Z 
```
But  we observe these lines after some time:
`Mar  2 20:56:16 lxnmspromscalev2p promscale[66744]: level=info ts=2022-03-02T19:56:16.579Z caller=throughput.go:76 msg="ingestor throughput" samples/sec=0 metrics-max-sent-ts=2022-03-02T17:32:13.673Z `

promscale keeps logging **samples/sec=0** which indicates that ingestion is not working normally.
Promscale will eventually eat up all of the memory and system will terminate it.

 We followed recommendations from command:
` timescaledb-tune -dry-run -conf-path /app/pgsql/13/data/postgresql.conf -max-conns 1000 -memory 100GB -quiet -pg-config /usr/pgsql-13/bin/pg_config
`
Can you propose any procedures on what to check or tune on the promscale/timescale db?

Thanks in advance!

**Comment by JamesGuthrie** [comment]
@brasno Thanks for the bug report.

Can you give us an indication of the volume of metrics that you're trying to ingest (e.g. 100'000/s). Can you also see if there is anything interesting in the postgres/timscaledb logs?

**Comment by niksajakovljevic** [comment]
@brasno Do you mind sharing Prometheus configuration, in particular the `remote_writer` part. Also sharing Prometheus and TimescaleDB/Postgres logs would be useful. 
Can you please be more specific and tell what ingest numbers you had before and what are you seeing now?
From Promscale logs it looks like there is no data being ingested which is strange. What are the logs on another Promscale instance since you did mention running HA setup?
How does single instance of Promscale perform?

**Comment by hvulin** [comment]
I'll give some update until brasno has time... We are currently running 11 pairs of prometheus server over 2 promscale servers. Before "upgrade" of promscale we were over 100k/s ingestion but it was also overloaded and it would slowly drop over time while promscales started using more and more memory until OOM killer killed them. After upgrade to 0.10.0 ingestion is VERY slow:
Mar  3 10:00:30 lxnmspromscalev2p promscale[76132]: level=warn ts=2022-03-03T09:00:30.081Z caller=watcher.go:101 msg="[WARNING] Ingestion is a very long time" duration=25m55.468800507s threshold=1m0s

so we mostly get 0 samples/sec throughtput and from time to time (every couple of minutes) there shows one with between 10k-20k samples/s. Database is CPU overloaded all sessions use 20%CPU each, total i over 98%. It's not memory bound and there are no waiting for I/Os.

This is a sample of remote_write:
remote_write:
  - url: "http://192.168.250.86:9201/write"
    remote_timeout: 30s
    queue_config:
      capacity: 10000
      max_samples_per_send: 3000
      batch_send_deadline: 10s
      min_shards: 4
      #max_shards: 10
      max_shards: 200
      #min_backoff: 10s
      min_backoff: 100ms
      #max_backoff: 30s
      max_backoff: 10s

The only thing I currently see in master db log is strange since promscale shouldn't try writing to compressed data:
Mar  3 10:15:17 lxnmspromdbv2p postgres[564023]: [47-4] #011          SELECT * FROM unnest($1, $2, $3) a(t,v,s) ORDER BY s,t"
Mar  3 10:15:17 lxnmspromdbv2p postgres[564023]: [47-5] #011PL/pgSQL function insert_metric_row(name,timestamp with time zone[],double precision[],bigint[]) line 8 at EXECUTE
Mar  3 10:15:17 lxnmspromdbv2p postgres[564023]: [47-6] 2022-03-03 10:15:17.162 CET [564023] STATEMENT:  SELECT _prom_catalog.insert_metric_row($1, $2::TIMESTAMPTZ[], $3::DOUBLE PRECISION[], $4::BIGINT[])
Mar  3 10:15:17 lxnmspromdbv2p postgres[563504]: [115-1] 2022-03-03 10:15:17.196 CET [563504] ERROR:  insert/update/delete not permitted on chunk "_hyper_379_2786665_chunk"
Mar  3 10:15:17 lxnmspromdbv2p postgres[563504]: [115-2] 2022-03-03 10:15:17.196 CET [563504] HINT:  Make sure the chunk is not compressed.
Mar  3 10:15:17 lxnmspromdbv2p postgres[563504]: [115-3] 2022-03-03 10:15:17.196 CET [563504] CONTEXT:  SQL statement "INSERT INTO  prom_data.node_network_receive_multicast_total (time, value, series_id)

So there is ingestion but it's VERY slow...
This is sample of top from master:
Tasks: 761 total, 194 running, 567 sleeping,   0 stopped,   0 zombie
%Cpu(s): 98.7 us,  0.8 sy,  0.0 ni,  0.0 id,  0.0 wa,  0.4 hi,  0.1 si,  0.0 st
MiB Mem : 193168.6 total,    990.1 free,  22430.4 used, 169748.1 buff/cache
MiB Swap:  65536.0 total,  65471.0 free,     65.0 used. 140547.5 avail Mem

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 563579 postgres  20   0   28.0g   3.8g   3.8g R  33.3   2.0   6:19.67 postgres
 563509 postgres  20   0   28.0g   3.9g   3.9g R  25.0   2.1   7:35.34 postgres
 563548 postgres  20   0   28.0g   3.4g   3.3g R  25.0   1.8   7:46.82 postgres
 563480 postgres  20   0   28.0g   3.5g   3.5g R  20.8   1.9   8:07.92 postgres
 563492 postgres  20   0   28.0g   4.6g   4.5g R  20.8   2.4   7:07.68 postgres
 563493 postgres  20   0   28.0g   2.1g   2.1g R  20.8   1.1   8:09.42 postgres
 563499 postgres  20   0   28.0g   4.8g   4.7g R  20.8   2.5   5:55.35 postgres
 563501 postgres  20   0   28.0g   3.8g   3.8g R  20.8   2.0   8:02.58 postgres
 563506 postgres  20   0   28.0g   2.8g   2.7g R  20.8   1.5   8:11.22 postgres
 563573 postgres  20   0   28.0g   3.4g   3.3g R  20.8   1.8   7:01.86 postgres

**Comment by hvulin** [comment]
this is from before the upgrade:
Feb 28 23:10:02 lxnmspromscalev1p promscale[476246]: level=info ts=2022-02-28T22:10:02.471Z caller=throughput.go:61 msg="Samples write throughput" samples/sec=99000 metadata/sec=0 max-sent-ts=2022-02-28T22:10
:01.394Z
Feb 28 23:10:03 lxnmspromscalev1p promscale[476246]: level=info ts=2022-02-28T22:10:03.470Z caller=throughput.go:61 msg="Samples write throughput" samples/sec=106085 metadata/sec=0 max-sent-ts=2022-02-28T22:1
0:02.548Z
Feb 28 23:10:04 lxnmspromscalev1p promscale[476246]: level=info ts=2022-02-28T22:10:04.471Z caller=throughput.go:61 msg="Samples write throughput" samples/sec=24000 metadata/sec=0 max-sent-ts=2022-02-28T22:10
:03.334Z
Feb 28 23:10:05 lxnmspromscalev1p promscale[476246]: level=info ts=2022-02-28T22:10:05.470Z caller=throughput.go:61 msg="Samples write throughput" samples/sec=112851 metadata/sec=0 max-sent-ts=2022-02-28T22:1
0:04.635Z
Feb 28 23:10:06 lxnmspromscalev1p promscale[476246]: level=info ts=2022-02-28T22:10:06.471Z caller=throughput.go:61 msg="Samples write throughput" samples/sec=44195 metadata/sec=0 max-sent-ts=2022-02-28T22:10
:05.848Z
Feb 28 23:10:07 lxnmspromscalev1p promscale[476246]: level=info ts=2022-02-28T22:10:07.471Z caller=throughput.go:61 msg="Samples write throughput" samples/sec=36000 metadata/sec=0 max-sent-ts=2022-02-28T22:10
:06.583Z
Feb 28 23:10:08 lxnmspromscalev1p promscale[476246]: level=info ts=2022-02-28T22:10:08.471Z caller=throughput.go:61 msg="Samples write throughput" samples/sec=22317 metadata/sec=0 max-sent-ts=2022-02-28T22:10
:07.765Z
Feb 28 23:10:09 lxnmspromscalev1p promscale[476246]: level=info ts=2022-02-28T22:10:09.470Z caller=throughput.go:61 msg="Samples write throughput" samples/sec=9000 metadata/sec=0 max-sent-ts=2022-02-28T22:10:
08.547Z
Feb 28 23:10:10 lxnmspromscalev1p promscale[476246]: level=info ts=2022-02-28T22:10:10.471Z caller=throughput.go:61 msg="Samples write throughput" samples/sec=43293 metadata/sec=0 max-sent-ts=2022-02-28T22:10
:09.222Z

**Comment by hvulin** [comment]
is there a way to downgrade promscale back to where it worked "ok" without dropping the DB and starting from scratch?

**Comment by hvulin** [comment]
update:
we created a new db cluster (again 32Core, 196G RAM each) with latest postgres 13 (13.6), latest timescaledb (2.6), and latest promscale (0.10.0) with a fresh database and this is now working fine (we had over 1.5M samples/s) - but the database is small... We then upgraded the old DB to 13.6 and timescale to 2.6 (it would be nice to update the documentation with expectation about how long this takes - on our 450G+ db alter extension took more then 30min). Then we tried again but it's still stuck... We get 2 maintenance jobs (probably doing the recompression of EVERYTHING since you had a fix for that) at 100% cpu each and ingestion is still slow (we tried with a single prometheus). Log said the usual "slow ingestion" message and it was 3.5min... 
Can you tell us WHAT is that recompress fix doing and will it have to recompress the whole database???? I presume that while this is going on something important is locked in the database and there is no ingestion until maint. job switches from chunk to chunk...

Is the 0.5 terabytes really such a big database that promscale can't handle it????

**Comment by Harkishen-Singh** [comment]
Hey @hvulin,

Can you update your prometheus remote-wrote config to below and see if it gives any improvement with 0.10. I feel that too many small HTTP requests may not be efficient.

```yaml
url: "http://192.168.250.86:9201/write"
remote_timeout: 5m
queue_config:
  capacity: 100000
  max_samples_per_send: 30000
  batch_send_deadline: 10s
  min_shards: 8
  max_shards: 2000
  min_backoff: 100ms
  max_backoff: 10s
```

> recompress fix doing 

What do you mean by compression? The maintenance jobs only compress the uncompressed chunks, not compressing again the compressed ones.

I think 2 maintenance jobs may not be sufficient for the data you have, which is why when the compression starts, it takes a lot of time since the 2 jobs have a lot of chunks to be compressed. 

try adding more 3 jobs by executing `PERFORM public.add_job('_prom_catalog.execute_maintenance_job', '30 min');`  3 times.

For the time, to decrease the database size, you can manually compress the database with `call prom_api.execute_maintenance();`. This may take sometime since you have a lot of uncompressed chunks, but it should help decrease the database size.

**Comment by hvulin** [comment]
we will try your suggestions and reply... as far as compression I'm talking about a earlier fix (we upgraded from 0.52 and there was a compression fix in the meantime) that had to fix an issue which left a single chunk uncompressed...

**Comment by niksajakovljevic** [comment]
@hvulin 

I'd recommend tweaking number of shards for `remote_write` configuration:
```
min_shards: 20
max_shards: 20
```
We've seen performance degradation of Promscale due to Prometheus resharding. The rest of your
`remote_write` configuration seems fine although you might want to tweak `capacity` to be around 10X of batch size `max_samples_per_send`.

**Comment by hvulin** [comment]
Hm, @Harkishen-Singh told us some other options but it doesn't really matter... Our new DB is coping for now but our old DB is still unresponsive after upgrade... Performance problem with promscale comes from unusable DB not from the prometheus side... We had 1 prometheus sending data to old db and it was still taking over 3min to ingest... Database isn't doing anything but maintenance jobs of compressing (we added 3 more jobs so now there are 5). Could this be related to the bug (https://github.com/timescale/promscale/pull/1081) and the fact that the DB needs to go back through almost all of the 450G of data and fix compression???? Nothing else comes to mind but ingestion is extremely slow on this old DB and works normaly on the new one (we configured couple of prometheuses to push data to both)

**Comment by brasno** [comment]
We changed **remote write** settings on prometheus according to previous recommendations for **capacity** (10000 before), **min_shards** (4 before) and **max_shards** (200 before):
```
remote_write:
  - url: "http://192.168.255.85:9201/write"
    remote_timeout: 30s
    queue_config:
      capacity: 30000
      max_samples_per_send: 3000
      batch_send_deadline: 10s
      min_shards: 20
      max_shards: 20
      min_backoff: 100ms
      max_backoff: 10s
```
After change, we observed increased ingestion time on promscales (rate(promscale_ingest_duration_seconds_count[5m]) of about 50%. Also database (32 cores) is now constantly over 90% CPU utilization.

**Comment by hvulin** [comment]
guys did some testing to see why the database is so slow even with small number of samples and here are some of the results..
![photo_2022-03-10_20-57-54](https://user-images.githubusercontent.com/60874879/157750190-c169c3d1-a9c0-4712-8b88-acfe85d3d81a.jpg)
![photo_2022-03-10_20-58-52](https://user-images.githubusercontent.com/60874879/157750280-0d35f7a7-9c89-4869-9415-e6f34829a7c8.jpg)

**Comment by brasno** [comment]
Dear all,
we are running two instances of promscale, but one died with out of memory three days ago. Second promscale took all load, and after some time, stabilized with approximately doubled RAM consumption. Here is RAM of **second promscale** last 7 days:
![image](https://user-images.githubusercontent.com/36503512/158150018-92f8e5d1-d453-4563-871e-f0055bd094f8.png)


**Timescaledb** machine also reduced CPU load from 95% to 45% some time after first promscale crash. CPU in last 7 days:
![image](https://user-images.githubusercontent.com/36503512/158149552-d03a14b3-1efc-4324-a5ab-e623aa54d2e3.png)

Obviously, for some reason if two promscales attack same timescaledb we experience heavy CPU load on timescaledb and consequently slower metric ingestion.
Can you please advise how to better scale promscales? Thanks.

**Comment by antekresic** [comment]
Hi there,

I've been going through this issue to get all the information regarding the performance issues you have been noticing lately. Specifically, from your last comment, it really looks like you have hit some sort of bug which is causing higher than usual resource usage.

In order to confirm this, I want double-check some facts about your setup (which isn't as trivial).

You are running multiple (11) Prometheus instances which you are pointing to both of your Promscale instances which are running in high availability mode. Promscale instances are pointing to a proxy of your high availability Postgres cluster managed by Patroni.

Is this correct? Are you distributing the load evenly over both of the Promscale instances?

**Comment by antekresic** [comment]
So looking at all of this, the RAM increase makes sense since one Promscale is taking on all of the write operations and doubling its cache size. What's not clear to me is why the CPU usage on the database dropped after one of the Promscale instances stopped working.

Did you try bringing up the second instance of Promscale back up again and did the DB CPU usage jump up again after that?

**Comment by mjandjel** [comment]
Deal all,
      we did try to let the second Promscale to write to the db and the symptoms are the same as already described by brasno:  
_"Obviously, for some reason if two promscales attack same timescaledb we experience heavy CPU load on timescaledb and consequently slower metric ingestion.
Can you please advise how to better scale promscales? Thanks."_

We still suffer from severe performance degradation on database side if more than one Promscale is writing to the same Timescale DB. Here is a preview of a performance chart for the Timescale node. At around 12:52 when we directed one of the Prometheus instances from promscale2 to promscale1 which before was only on standby, and the CPU on Timescale host went to 100%. 
![postgres_nodeexporter_dashboard](https://user-images.githubusercontent.com/23310076/158393277-fc0c8abb-2ce3-4d45-8754-39019c6242d0.png)

**Comment by hvulin** [comment]
CPU drop on DB with only 1 promscale is mostly "normal" since now DB has to handle 100 connections, while it had to handle 200 connections while there were 2 promscales... 1 promscale is now handling the combined load (number of ingested shards is mostly the same as promscale1+promscale2) but using half of the connections.... There seems to be some issue withing prom_api while higher parallelism (200 connections) is used compared to single promscale (100 connections) - I don't know if it's some sort of locking problem since I don't really know what is the expected behaviour....

**Comment by antekresic** [comment]
OK, there is one thing there that you need to realize wrt DB CPU usage on Promscale startup. When you startup an instance, its series cache needs to be populated which can be very CPU intensive on the database. However, this is a temporary slowdown which should resolve itself once the cache has been populated. So, expected behavior for now and something we are looking into improving in the future.

My suggestion is to try this pointing the load promscale2 instance again and give it a bit of time while it populates the cache. After a while (which depends on the amount of active series you are ingesting), performance should improve and DB CPU should drop to the expected 45% like with promscale1. After that, splitting the load over both of them should have the same resource requirements like when running a single instance. 

Do let me know if this makes sense and works out for you.

**Comment by antekresic** [comment]
I see... I'm sorry for the long back and forth here, I'm just trying to understand why is this happening. 

I only just noticed that you have the DB connection concurrency flag set to 12. That actually might be the root of the problem here. Promscale will startup multiple concurrent writers based on your machines CPU core count. If you set that flag to 12, it will start that count times 12. I doubt that is what you want since you have a 32 core CPU on the DB. 

Could you try reducing that to 1 or 2 (depending on the amount of cores you have on Promscale boxes)? For instance, if you are running on 12 core CPUs, 1 should be pretty good, so you get 24 writers writing to the DB at the same time from both Promscale instances. If you are running on a 6 core CPU, I would use 2 etc.

I'm hoping this reduces the amount of lock contention in the database and CPU usage.

**Comment by brasno** [comment]
Dear @antekresic  I just wanted to write you that we reduced our promscales to only one because we experienced 99,9% CPU utilization on DB box. We have 8 cores on each promscale, so we'll try with  concurrency flag set to 1.

**Comment by antekresic** [comment]
Awesome, feel free to let me know of the outcome of your change.

**Comment by brasno** [comment]
Dear @antekresic , we tried promscale1 and 2 with PROMSCALE_DB_WRITER_CONNECTION_CONCURRENCY set to:
1,1; 2,2; 
We excluded promeheus pairs, ie. only one is sending samples to promscales.
Here is graph with **Samples ingestion** annotated with points in time when we changed parameters. Unfortunately, we cannot work with two promscales in parallel. Obviously something was changed from V0.5.2 to V0.10.0. which prevents us to work in that way. At the end, we changed parameter PROMSCALE_DB_WRITER_CONNECTION_CONCURRENCY=4 and continue with only one promscale. 

![image](https://user-images.githubusercontent.com/36503512/159328792-67ac5961-bab0-4fc1-9183-070b6ae58ae7.png)

Reducing concurrency parameter to 4 (assuming 4x8 cores =32 - number of cores on db machine)  greatly reduced rowexclusivelocks. Here is pic of it in the same time period as on previous pic. Database CPU utilization is now around 90%. 

![image](https://user-images.githubusercontent.com/36503512/159331266-be86f2c0-a7d3-4ecb-ba7d-ff1c6171616a.png)
 
Any idea what we have to do to be able to use two or more promscales? Thanks!

**Comment by antekresic** [comment]
Thanks for the update.

Just to confirm once again, the main issue is DB CPU utilization with the latest version when running multiple Promscale instances simultaneously? And you can confirm that the issue wasn't there in version 0.5.2?

Will look into this a bit more and get back to you.

**Comment by brasno** [comment]
Yes. We had version 0.5.2 running, but we read how many things were fixed in version 0.10.0 and decided to upgrade because we were planning to add more exporters. Unfortunately, it didn't work well, and we made a completely new installation with a completely empty database. After @niksajakovljevic asked how single instance works, we saw that it works better than two instances.
We have made progress, but we are still experiencing great CPU utilization of the database and message msg = "[WARNING] Ingestion is a very long time" in the promscale log with increasing values, and after a while we have to restart both the database and promscale.

**Comment by VineethReddy02** [comment]
Hi @brasno @hvulin @mjandjel 

I see you are facing issues with scaling the platform post upgrades. Do you mind joining our [Timescale community slack]([http://tsdb.co/community-slack](https://t.co/0FA2NXNG2g)) to start a follow-up discussion on this and how can the team better assist issues on your end? If required we can also jump on to a call to discuss the issues and to help you in succeeding with Promscale. Having you folks in our community slack eases the discussion on your issues. 

If this sounds good drop me a message in our Timescale community slack or confirm me here once you sign up in TimescaleDB slack will drop you a message. Looking forward to supporting you folks with a smooth Promscale experience! :)

**Comment by brasno** [comment]
Hi @VineethReddy02, 
glad to hear you! I joined slack, and others will join soon (tomorrow). I believe you can guide us through the channels to reduce noise to a minimum and quickly find a solution to our challenges. CU

**Comment by VineethReddy02** [comment]
@brasno The Promscale team is active in #promscale channel. You can find me in slack with`Vineeth` (followed by the Eon logo representing the Timescale team). I just dropped you (brasno in Timescale slack) a message.

**Comment by hvulin** [comment]
Latest update... We had a session yesterday that seems to have helped a lot...
On prometheus servers (all 22) we adjusted: 
remote_timeout: 60s
min_shards: 20
max_shards: 20
capacity: 200000
max_samples_per_send: 20000
which probably helped some but not with the database...
We fixed the promscale config (and run with 1 promscale):
PROMSCALE_DB_CONNECTIONS_MAX="50"
PROMSCALE_DB_WRITER_CONNECTION_CONCURRENCY=2
and removed -metrics.async-acks=true from startup options but fixed caches -metrics.cache.metrics.size=10000000 -metrics.cache.series.initial-size=10000000
this also helped but not as much the change on the database...
On the database we used patroni edit-config and added synchronous_commit: false and then restarted the database...

This final change made the database breathe again... now our CPU usage around 30% with peaks to 55%...
We still need to test with all prometheus servers (not just 1 in each cluster) and with both promscales behind a load balancer...
I presume we will have to change max_connections to 25 and concurrency to 1 on both but we will see next week.

Thanks for your help people.

**Comment by mjandjel** [comment]
Hi Team!

For the past one week the performance of the system appeared to be stable. Database CPU load was significantly lower, around 50% in average.
Now we are facing different performance degradation which manifest as oscillating instabilities lasting for several hours, occurring at random times.
The basic symptoms are:
	
	- Prometheuses reporting msg="Failed to send batch, retrying" err="Post \"http://192.168.250.86:9201/write\": context deadline exceeded"
	- Promscales reporting  msg="[WARNING] Ingestion is a very long time" duration=3m13.291186566s and msg="duplicate data in sample" total-duplicate-metrics=3533
	- Promscale metric reporting Prometheus leader switching
	- Postgres machine increased CPU load to 80-90 percent (vs 40-50% CPU load when working ok)
	- Postgres logs showing ERROR:  LEADER_HAS_CHANGED
	- Postgres metric shows increased number of accessharelock ( a couple of hundreds instead of up to one hundred)

In order to mitigate the issue we tried to increase ha leases parameters in timescale:
	ha_lease_refresh 10s -> changed to -> 30s
	ha_lease_timeout 1m -> changed to -> 5m
	
![image](https://user-images.githubusercontent.com/23310076/161935847-2c388ec0-4a9a-478a-a2ae-3bce80d5ccac.png)
![image](https://user-images.githubusercontent.com/23310076/161935941-42dc91aa-f0aa-44f1-8411-49d2a5ab807e.png)

We did not restart the Promscales or Prometheuses so we were not sure if these settings applied or not.
However, we did not noticed any effect. The Prometheuses would still change leaders when performance oscitations occurred. 

We are in process of shutting down one Promscale instance and running only on a single Promscale instance, hoping that this will alleviate symptoms as we suspect that for some reason, our system cannot run stable with 11 Prometheus clusters with two Promscales. We are also increasing the number of maintenance jobs from 2 to 8, as recommended by the Promscale team [#1282] (https://github.com/timescale/promscale/issues/1282)	

We also had an episode of Postgres transaction wraparound issue on 1st April but that is another issue probably related to autovacuuming not able to complete its jobs. 

Does the Promscale team have any recommendation on tuning the mentioned Prometheus HA leases parameters for our setup?  Did we had to restart the system for the change to take effect? Any clues or hints at where to look at to get to the source of the problem?
